Unmarshal "allOf" with non-object type sub-schemas

sisp opened the pull request:
This PR adds support for two related aspects of unmarshaling data using an OpenAPI schema:

1. Specifying an array schema _without_ `items` is valid, but `python-oas` currently requires the presence of `items`. This PR returns the raw array instance without further unmarshaling when `items` is not present. Specifying an array schema without `items` can be useful, e.g., for adding constraints such as `{min,max}Items` to another array schema using `allOf`.
1. `python-oas` currently only supports object sub-schemas of `allOf`. This PR adds support for other sub-schemas, in particular array sub-schemas, with the limitation that the first sub-schema is used for unmarshaling. There are a few edge cases where this behavior results in slightly incorrect unmarshaling, e.g. when the first sub-schema is an array schema _without_ `items` and the second sub-schema is an array schema _with_ `items` with date values that should be parsed. The proper solution to this problem is to merge all sub-schemas of `allOf`, but it seems not trivial (see e.g. [`json-schema-merge-allof`](https://github.com/mokkabonna/json-schema-merge-allof)). I couldn't find a Python library for merging OpenAPI schemas, so we would need to write. I think having support for non-object sub-schemas with some limitations is better than not supporting non-object sub-schemas at all for now.

Commit:
Unmarshal arrays without "items" property

Commit:
Unmarshal "allOf" with non-object sub-schemas

grktsh commented:
Thanks but I think that specifying an array schema without items is invalid. [OpenAPI](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#properties) says as follows:

> items - Value MUST be an object and not an array. Inline or referenced schema MUST be of a Schema Object and not a standard JSON Schema. *items MUST be present if the type is array*.

Commit:
Revert 'Unmarshal arrays without "items" property'

This reverts commit f45ce82b17249a278edda90d0cfef53bde49911e.

Commit:
Remove tests for "allOf" with arrays without "items" property

sisp commented:
Oh, you're right, I didn't realize this decrepancy between OpenAPI and JSON Schema. Thanks for pointing it out! Instead of omitting `items`, I can simply set `"items": {}`.

sisp commented:
I have reverted f45ce82b17249a278edda90d0cfef53bde49911e and removed the tests for `allOf` with array sub-schema in which `items` is not present. Then, this PR still adds support for non-object sub-schemas of `allOf`.

What do you think?

Commit:
Add test that unmarshaling with "allOf" and array sub-schemas validates array constraints

Commit:
Add tests for unmarshaling with "allOf" and sub-schemas with primitive types

grktsh reviewed (commented):
`"items": {}` is good! I left a minor comment.

Commit:
Check the type property of the first allOf sub-schema instead of checking the unmarshaled instance type

sisp commented on src/oas/schema/unmarshalers.py:
```diff
@@ -39,16 +41,21 @@ def _unmarshal(self, instance, schema):
+            result = self._unmarshal(instance, next(sub_schemas))
+            # If the first sub-schema of ``allOf`` unmarshals a ``dict``-like
+            # object, also unmarshal with the remaining sub-schemas.
+            if isinstance(result, Mapping):
```
Good point. I've changed the check as you suggested.

grktsh commented:
Thanks, merged!
